source/adapters/opencl/usm.cpp

```diff
@@ -343,18 +404,59 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMFill2D(
     [[maybe_unused]] uint32_t numEventsInWaitList,
     [[maybe_unused]] const ur_event_handle_t *phEventWaitList,
     [[maybe_unused]] ur_event_handle_t *phEvent) {
-  return UR_RESULT_ERROR_INVALID_OPERATION;
+  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMMemcpy2D(
-    [[maybe_unused]] ur_queue_handle_t hQueue, [[maybe_unused]] bool blocking,
-    [[maybe_unused]] void *pDst, [[maybe_unused]] size_t dstPitch,
-    [[maybe_unused]] const void *pSrc, [[maybe_unused]] size_t srcPitch,
-    [[maybe_unused]] size_t width, [[maybe_unused]] size_t height,
-    [[maybe_unused]] uint32_t numEventsInWaitList,
-    [[maybe_unused]] const ur_event_handle_t *phEventWaitList,
-    [[maybe_unused]] ur_event_handle_t *phEvent) {
-  return UR_RESULT_ERROR_INVALID_OPERATION;
+    ur_queue_handle_t hQueue, bool blocking, void *pDst, size_t dstPitch,
+    const void *pSrc, size_t srcPitch, size_t width, size_t height,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+  cl_context CLContext;
+  CL_RETURN_ON_FAILURE(clGetCommandQueueInfo(
+      cl_adapter::cast<cl_command_queue>(hQueue), CL_QUEUE_CONTEXT,
+      sizeof(cl_context), &CLContext, nullptr));
+
+  clEnqueueMemcpyINTEL_fn FuncPtr = nullptr;
+  ur_result_t RetVal = cl_ext::getExtFuncFromContext<clEnqueueMemcpyINTEL_fn>(
+      CLContext, cl_ext::ExtFuncPtrCache->clEnqueueMemcpyINTELCache,
+      cl_ext::EnqueueMemcpyName, &FuncPtr);
+
+  if (!FuncPtr) {
+    return RetVal;
+  }
+
+  std::vector<cl_event> Events;
```

Comment:
Small performance potential improvement could be to reserve the vector memory in advance instead of relying to pull_back to resize it